### ISSUE TITLE
[feature request] allow DISTINCT function to operate on tags

### ISSUE THREAD

**Issue by The-Nik** [influxdata/influxdb#3880]
I would like to have following feature:

Sience the new release 0.9.3 tags are resulting as own columns if you use SELECT * FROM measurement

Currently, it's not possible to use commands on this columns. An example:

    SELECT * FROM measurements

returns:

    time tagA tagB value
    xxx M N 0.3
    xxx M O 0.4
    xxx P R 0.2

I want to do a query like:

    SELECT count(distinct(tagA)) FROM measurements

The result is

    2 (M+P)

Anyone else need this feature?

**Comment by The-Nik** [comment]
Idea also exist in #1815 at "Not currently implemented (might in the future, but no promises)"

**Comment by beckettsean** [comment]
@The-Nik are you asking for `DISTINCT` to support tags, or are you asking for the same functionality that `SELECT *` used to do? To get the similar `SELECT *` behavior, just include a `GROUP BY *`

**Comment by The-Nik** [comment]
I ask for DISTINCT to support tags for counting the different tag values :+1:

**Comment by beckettsean** [comment]
@The-Nik you can use `SHOW TAG VALUES` plus some shell to get what you want:

`$ influx -execute 'show tag values with key = a' -database mydb` will print a list of all tag values associated with the key `a` on the database `mydb`. The output has two header lines, so if you pass it to `wc -l` just subtract 2 for the actual count:

`$ influx -execute 'show tag values with key = a' -database mydb | wc -l` and then subtract 2 from the output.

**Comment by The-Nik** [comment]
Yeah, this is a good way. But in my case, I need the number in Grafana in a single stat panel. In Grafana, there are some aggregate fuctions but no "count". So InfluxDB has to serve the exact value or I build something in my Grafana like a count-function ;-)

**Comment by yee379** [comment]
+1; being able to quickly summarise the distinct number of datasets / tags directly from the influx SQL would be very handy; e.g. a grafana panel of the number of sensors I have reporting data over time.

**Comment by edennis-sge** [comment]
+1: I have a similar use case to yee379 in mind.

**Comment by jakefoster** [comment]
+1 on being able to count distinct tags.

I also feels like this speaks to the deeper issue of providing guidance on what should be a tag vs. a value.  For a schema-less DB there's sure a lot of subtlety around defining your schema!  :)

**Comment by morganda** [comment]
I would also like to see this. We use the cpu and load plugins which themselves don't explicitly provide the cpu count. They do provide each cpu as an "instance" tag e.g. a box with 32 cpus will collect metrics on 32 individual cpus tagging them with with their instance number. If I could get the total count of cpus from the tags, then the load numbers would have a little more context in our graphs Grafana and Chronograph.

**Comment by JulienChampseix** [comment]
+1, any progress on it ?

**Comment by RobertAtomic** [comment]
+1. this would makes things quite a bit simpler for some tasks at hand.

**Comment by rafael84** [comment]
+1

**Comment by ohmystack** [comment]
+1 Really need this. Tag should also support a kind of normal `SELECT` search, which can be handled by Grafana.

**Comment by thepolina** [comment]
+1
Desperately need this

**Comment by Guibod** [comment]
+ 1
Anybody as a solution to count my hosts in Grafana through Influxdb query language ?

**Comment by tomhallam** [comment]
+1

**Comment by selzoc** [comment]
+1

**Comment by mosoto** [comment]
+1

**Comment by davidgardner11** [comment]
++++++1 This would really help pulling some of our metrics much much easier

**Comment by brumfb** [comment]
+1

**Comment by beckettsean** [comment]
Perhaps a better way to accomplish the same goals: https://github.com/influxdata/influxdb/issues/5668

**Comment by supershal** [comment]
+1

**Comment by matt-snider** [comment]
+1

**Comment by cmasekera** [comment]
+1

**Comment by gvohra** [comment]
+1

**Comment by cnelissen** [comment]
+1

**Comment by jwestboston** [comment]
+1

**Comment by gabtastic** [comment]
+1

**Comment by skv2602** [comment]
+1

**Comment by drewdavies** [comment]
+1

**Comment by remipannequin** [comment]
+1
The use case where I need this feature : to monitor the number of sources (in my case wireless temperature sensors) that have submitted at least one value in -say- the last 30 minutes, i.e. that are alive.

**Comment by bricsuc** [comment]
+1. You can fake this feature by duplicating the tag value in a field, but that's a pretty bogus solution.

**Comment by bertpig** [comment]
+1

**Comment by jckwon** [comment]
+1

**Comment by saboteurkid** [comment]
+1

**Comment by gregorg** [comment]
+1

**Comment by smeapng** [comment]
+1

**Comment by J-Mx** [comment]
+1 we really need it !

**Comment by erowan** [comment]
+1

**Comment by sachinrase** [comment]
+1  , it would be really helpful.

**Comment by bmundt** [comment]
+1

**Comment by andyxning** [comment]
+1  @beckettsean any progress on this feature?

**Comment by xo4n** [comment]
+1

**Comment by bruce-jin** [comment]
+1

**Comment by cloudnull** [comment]
+1

**Comment by donnut** [comment]
+1

**Comment by jsternberg** [comment]
We're going to implement this with #7195.

**Comment by cnelissen** [comment]
@jsternberg I don't think the proposal from #7195 accurately addresses the issue here... The desired fix (highlighted in this issue) should be part of the SELECT syntax, not an entirely new query.

For instance, I have measurements that use device hardware addresses as the tag, and I want to get a count of all unique devices along with averages of a few fields. For instance, if I wanted to get the total count of connected wireless clients and how much bandwidth was used in the last day in 5 minute intervals:

```
SELECT
    COUNT(DISTINCT(macAddress)), <-- This is a tag
    SUM(downloadBytes),
    SUM(uploadBytes)
FROM "wirelessClients"
WHERE time > NOW() - 1d
GROUP BY time(5m)
```

This is what we are hoping to achieve. The proposal in #7195 will not address this use case.

**Comment by jsternberg** [comment]
I'm going to reopen this and think about it over the weekend. I'm still not sure we can do what you're requesting, but it sounds different enough from `SHOW CARDINALITY` that it deserves some discussion about the issue before closing.

**Comment by yoyomikeyc** [comment]
This would be valuable. +1

**Comment by dineshvenkat** [comment]
+1

**Comment by ahvetskovich** [comment]
+1

**Comment by Tiinusen** [comment]
+1

**Comment by jregovic** [comment]
+1

**Comment by stevenh** [comment]
+1

**Comment by aaskey** [comment]
+1

**Comment by laggyone** [comment]
+1

**Comment by gloyka** [comment]
+1

**Comment by bewja** [comment]
+1.  This point is a huge short coming, need to be able to select discinct on tags

**Comment by biker73** [comment]
+1

**Comment by leforsman** [comment]
+1

**Comment by strongpauly** [comment]
+1 Definitely a limiting factor on one of our key usecases.

**Comment by lobando** [comment]
+1

**Comment by bbala-github** [comment]
+1

**Comment by orangle** [comment]
+1

**Comment by damarnez** [comment]
+1

**Comment by SimSimY** [comment]
+1 (as described by @cnelissen)

**Comment by lucadistefano** [comment]
+1

**Comment by JamesClonk** [comment]
+1

**Comment by jsternberg** [comment]
Please leave +1 comments to adding a 👍 to the top post using a reaction. Leaving a message notifies everybody who is participating in this conversation and doesn't add anything to the discussion.

**Comment by sidgod** [comment]
+1

**Comment by joriws** [comment]
My need is to count number of unique tags with tag filters in Grafana. I can count fields but it gives incorrect answer. SHOW SERIES cannot be limited enough like return only one TAG which I could the distinct + count.

SELECT count("Incoming_Answers_2xxx") FROM "Realm-day" WHERE "INSTANCE" =~ /IPXDEA/ AND "ANSWERHOST" =~ /dtag/ AND "REALM" =~ /mcc2/ AND time > '2017-02-09T00:00:00Z';
name: Realm-day
---------------
time                    count
1486598400000000001     204

There I would like to have
SELECT count(distict(REALM)) FROM "Realm-day" WHERE "INSTANCE" =~ /IPXDEA/ AND "ANSWERHOST" =~ /dtag/ AND "REALM" =~ /mcc2/ AND time > '2017-02-09T00:00:00Z';
name: Realm-day
---------------
time                    count
1486598400000000001     34

Or someting like  SHOW SERIES COUNT(DISTICT(TAG("REALM"))) FROM "Realm-day" WHERE "INSTANCE" =~ /IPXDEA/ AND "ANSWERHOST" =~ /dtag/ AND "REALM" =~ /mcc2/ AND time > '2017-02-09T00:00:00Z';

**Comment by lpic10** [comment]
I managed to achieve this by using subqueries in influxdb 1.2

Eg. getting number of hosts from telegraf in grafana:

`select count(tot) from (SELECT mean("used") as tot FROM "mem" WHERE $timeFilter GROUP BY "host" fill(null))`

I'm using a measurement and a field I know it will always be present, it could be anything.

**Comment by biker73** [comment]
If you get no data for a host for the time period won't it be missed ? I don't think this can be 100% relied upon ?

**Comment by lpic10** [comment]
Yes, but that's what I expected. If there is no data for a particular host during the selected query period I don't want to consider it. You can remove or maybe increase this time restriction in the WHERE clause, but then I guess the query can be quite slow.

**Comment by biker73** [comment]
Ok so it's slightly different use case, I think most want a distinct list of tag key values regardless of time period. 

i.e. I'd want all time across a year of data for example potentially peta bytes of data where the series count might be 2m cardinality.

**Comment by juddgaddie** [comment]
+1

**Comment by samjetski** [comment]
In my case I needed to display the number of sensors which reported within a time interval (to indicate confidence of the mean). I managed to work around it with a subquery, but it's a bit filthy:

```sql
SELECT count("first") FROM (
  SELECT first("value") FROM "mqtt_consumer"
  WHERE "topic" =~ /hub0[1234567]\/sensors\/\d+\/temperature/ AND $timeFilter
  GROUP BY time($interval), topic
)
WHERE $timeFilter
GROUP BY time($interval)
```

**Comment by lin-credible** [comment]
+1

**Comment by ampersand8** [comment]
+1

**Comment by jsternberg** [comment]
I'm locking this to prevent further 👍 messages. We will be discussing this to figure out the feasibility of the request and create a timeline. Please push the "Subscribe" button instead to get any updates about this feature.
